### PR TITLE
add new resource of api gateway api

### PR DESCRIPTION
**Pull request by geeklenny**
TF_ACC=1 go test ./alicloud -v -run=TestAccAlicloudApigatewayApi_basic -timeout=300m
=== RUN   TestAccAlicloudApigatewayApi_basic
b1604cf1a8b044e681dde9ecb527e388
--- PASS: TestAccAlicloudApigatewayApi_basic (2.13s)
PASS
ok  	github.com/terraform-providers/terraform-provider-alicloud/alicloud	(cached)

### CODE REVIEW

File: alicloud/resource_alicloud_api_gateway_api.go [hunk 1]
```diff
@@ -0,0 +1,808 @@
...
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"body_format": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+			},
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_api_gateway_api.go]: I think `request_config ` use TypeMap will be better.

File: alicloud/resource_alicloud_api_gateway_api.go [hunk 2]
```diff
@@ -0,0 +1,808 @@
...
+
+			"service_type": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"http_service_config": &schema.Schema{
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_api_gateway_api.go]: use typemap?

File: alicloud/resource_alicloud_api_gateway_api.go [hunk 3]
```diff
@@ -0,0 +1,808 @@
...
+						"aone_name": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+			},
+
+			"http_vpc_service_config": &schema.Schema{
+				Type:     schema.TypeList,
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_api_gateway_api.go]: use typemap?

File: alicloud/resource_alicloud_api_gateway_api.go [hunk 4]
```diff
@@ -0,0 +1,808 @@
...
+						"aone_name": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+			},
+
+			"mock_service_config": &schema.Schema{
+				Type:     schema.TypeList,
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_api_gateway_api.go]: use typemap?

File: alicloud/resource_alicloud_api_gateway_api.go [hunk 5]
```diff
@@ -0,0 +1,808 @@
...
+						"result": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+
+			"request_parameters": &schema.Schema{
+				Type:     schema.TypeSet,
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_api_gateway_api.go]: use typemap?

File: alicloud/resource_alicloud_api_gateway_api.go [hunk 6]
```diff
@@ -0,0 +1,808 @@
...
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+			},
+
+			"constant_parameters": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_api_gateway_api.go]: use typemap?

File: alicloud/resource_alicloud_api_gateway_api.go [hunk 7]
```diff
@@ -0,0 +1,808 @@
...
+						"description": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+			},
+
+			"system_parameters": &schema.Schema{
+				Type:     schema.TypeSet,
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_api_gateway_api.go]: use typemap?

File: website/docs/r/api_gateway_api.html.markdown [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+* `system_parameters` - (Required, Type: list) http_service_config defines the config when service_type selected 'HTTP'.
+    * `name` - (Required) System parameter name which supports values including in [system parameter list](https://www.alibabacloud.com/help/doc-detail/43677.html)
+    * `in` - (Required) System parameter location; values: 'HEAD' and 'QUERY'.
+    * `name_service` - (Required) Backend service's parameter name.
+
+## Attributes Reference
+
+The following attributes are exported:
+
+* `id` - The ID of the api of api gateway.
```
Comment (xiaozhu36) [inline website/docs/r/api_gateway_api.html.markdown]: If the resource support `Importer`, please show it in the docs.

**Comment by geeklenny** [comment]
all resolved

File: alicloud/resource_alicloud_api_gateway_api.go [hunk 9]
```diff
@@ -0,0 +1,814 @@
...
+	split := strings.Split(d.Id(), COLON_SEPARATED)
+	req.ApiId = split[0]
+	req.GroupId = split[1]
+	_, err := client.WithCloudApiClient(func(cloudApiClient *cloudapi.Client) (interface{}, error) {
+		return cloudApiClient.DeleteApi(req)
+	})
+	if err != nil {
+		return fmt.Errorf("Error deleting Api failed: %#v", err)
+	}
+	return nil
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_api_gateway_api.go]: Please ensure the api has been deleted after the deleting

File: alicloud/resource_alicloud_api_gateway_api.go [hunk 10]
```diff
@@ -0,0 +1,814 @@
...
+	req.RequestParameters = string(rps)
+	req.ServiceParameters = string(sps)
+	req.ServiceParametersMap = string(spm)
+
+	if update {
+		req.ResultType = "JSON"
+		req.ResultSample = "Result Sample"
+		req.Visibility = "PRIVATE"
+		req.AllowSignatureMethod = "HmacSHA256"
+		req.WebSocketApiType = "COMMON"
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_api_gateway_api.go]: Please use const parameters instead hard code

File: alicloud/resource_alicloud_api_gateway_api_test.go [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func init() {
+	resource.AddTestSweepers("alicloud_api_gateway_api", &resource.Sweeper{
+		Name: "alicloud_api_gateway_api",
+		F:    testSweepApiGatewayApi,
+		Dependencies: []string{
+			"alicloud_api_gateway_group",
+		},
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_api_gateway_api_test.go]: I think API gateway group sweeper has a dependence api gateway api

File: website/docs/r/api_gateway_api.html.markdown [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+Basic Usage
+
+```
+name = "terraformapi"
+group_id = "${alicloud_api_gateway_group.apiGatewayGroup.id}"
+description = "description"
+auth_type = "APP"
+request_config = [
+  {
+    protocol        = "HTTP"
```
Comment (xiaozhu36) [inline website/docs/r/api_gateway_api.html.markdown]: Please remove extra blank

File: website/docs/r/api_gateway_api.html.markdown [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+* `constant_parameters` - (Required, Type: list) http_service_config defines the config when service_type selected 'HTTP'.
+    * `name` - (Required) Constant parameter name.
+    * `in` - (Required) Constant parameter location; values: 'HEAD' and 'QUERY'.
+    * `value` - (Required) Constant parameter value.
+    * `description` - (Optional) The description of Constant parameter.
+* `system_parameters` - (Required, Type: list) http_service_config defines the config when service_type selected 'HTTP'.
+    * `name` - (Required) System parameter name which supports values including in [system parameter list](https://www.alibabacloud.com/help/doc-detail/43677.html)
+    * `in` - (Required) System parameter location; values: 'HEAD' and 'QUERY'.
+    * `name_service` - (Required) Backend service's parameter name.
+
```
Comment (xiaozhu36) [inline website/docs/r/api_gateway_api.html.markdown]: Please use blocker to show the complex configure

File: alicloud/resource_alicloud_api_gateway_api.go [hunk 14]
```diff
@@ -0,0 +1,814 @@
...
+	}
+
+	config := l[0].(map[string]interface{})
+	serviceConfig.Protocol = "HTTP"
+	serviceConfig.Address = config["address"].(string)
+	serviceConfig.Path = config["path"].(string)
+	serviceConfig.Method = config["method"].(string)
+	serviceConfig.Timeout = config["timeout"].(int)
+	serviceConfig.MockEnable = "FALSE"
+	serviceConfig.VpcEnable = "FALSE"
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_api_gateway_api.go]: Please use const

**Pull request merged**